Stop newline characters from accidentally being added when describing…

randompixel opened the pull request:
So one day when writing a description on my Macbook I accidentally hit Shift-Enter in the middle of the line (It's near " - what can I say?). 

I tried to fix it by carrying on....
<img width="178" alt="screen shot 2016-09-01 at 21 41 43" src="https://cloud.githubusercontent.com/assets/2316601/18183581/ee7d96b8-708c-11e6-946b-fe7c78acfe12.png">

However PHPSpec continues and you end up with something very magical and a bit broken:
<img width="243" alt="screen shot 2016-09-01 at 21 35 10" src="https://cloud.githubusercontent.com/assets/2316601/18183620/1db8b85e-708d-11e6-9b4d-b925fd7f4bac.png">

This patch is a simple one that will strip the OS default newline character out of the name. There should never be an instance when you'd want a new line in the middle of a describe string. This gives you:
<img width="314" alt="screen shot 2016-09-01 at 21 34 40" src="https://cloud.githubusercontent.com/assets/2316601/18183649/3d2c597a-708d-11e6-9283-e8ff72bccad5.png">

I did have a look around the tests to see if there was something to update but couldn't find anything. If you have any suggestions I'm happy to update this PR.

ciaranmcnulty commented:
This is definitely a bug, but I'd rather there was an error than a silent correction.

We currently check the class name is valid in PSR-0: https://github.com/phpspec/phpspec/blob/master/src/PhpSpec/Locator/PSR0/PSR0Locator.php#L360 so it'd rather catch here. It looks like the regex isn't multiline.

Would you be able to add the checks here perhaps?

randompixel commented:
I've updated the PR based on your suggestions. The regex seemed like the most sensible place to put the change (although _what_ a regex!). I also added a spec to cover the use-case.

I've tested on my Mac with a number of combinations of namespaces and line-breaks. Is this worth testing on a Windows machine too?

![screen shot 2016-09-13 at 21 34 43](https://cloud.githubusercontent.com/assets/2316601/18490470/22d253f6-79fa-11e6-824d-0e6ea6f17589.png)

ciaranmcnulty commented on src/PhpSpec/Locator/PSR0/PSR0Locator.php:
```diff
@@ -357,7 +357,7 @@ private function createResourceFromSpecFile($path)
     private function validatePsr0Classname($classname)
     {
-        $pattern = '/^([a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*[\/\\\\]?)*[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*$/';
+        $pattern = '/^[^\r\n]([a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*[\/\\\\]?)*[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*$/';
```
This isn't correct, you're allowing any char that isn't `\r` or `\n` to be the first character of the class name, including a bunch of invalid ones. I\'m not really clear why this makes your test pass, to be honest.

randompixel replied:
OK. Happy to rework.

ciaranmcnulty replied:
Ah ok it might be that `preg_match` is not in multiline mode so `^` and `$` are matching 'start of line' and 'end of line', that means extra linebreaks are allowed before and after.

You could try using `\A` and `\z` instead (note not `\Z` as that would allow a linebreak at the end), or try the `/m` or `/D` switches at the end of the pattern to affect the default behaviour of `^` and `$`.

randompixel replied:
Yup. I went with \z and updated the spec to better capture that.

randompixel commented:
After rebase and a bit more investigation it seems
<img width="632" alt="screen shot 2016-09-17 at 21 15 27" src="https://cloud.githubusercontent.com/assets/2316601/18610748/e43d0fc6-7d1b-11e6-98df-209ec4003e82.png">
was being blocked successfully but
<img width="471" alt="screen shot 2016-09-17 at 21 13 52" src="https://cloud.githubusercontent.com/assets/2316601/18610746/cc15bcea-7d1b-11e6-858d-5af62f763601.png">
was being let through. This updated regexp I hope should fix that issue.
<img width="637" alt="screen shot 2016-09-17 at 21 11 24" src="https://cloud.githubusercontent.com/assets/2316601/18610731/6bccf31c-7d1b-11e6-9461-9343d41bc551.png">

ciaranmcnulty commented on src/PhpSpec/Locator/PSR0/PSR0Locator.php:
```diff
@@ -357,7 +357,7 @@ private function createResourceFromSpecFile($path)
     private function validatePsr0Classname($classname)
     {
-        $pattern = '/^([a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*[\/\\\\]?)*[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*$/';
+        $pattern = '/^([a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*[\/\\\\]?)*[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*\z$/';
```
You don't need the dollar at the end if you've added `\z`. You can replace the `^` at the start with `\A` too

randompixel replied:
Updated as requested :)

ciaranmcnulty commented:
Great, thanks for this!

ciaranmcnulty commented:
Ah as this is a bug fix I should have merged this to the `2.5` branch - will fix

randompixel commented:
Thanks for the mentoring too 🙂
